Tradução: databases.qmd (#59)
cienciadedatos/pt-r4ds: databases.qmd

-   SGBDs na **Nuvem** (*Cloud*), como o *Snowflake*, *RedShift* da *Amazon* e o *BigQuery* do *Google*, são similares aos SGBDs cliente-servidor, mas eles rodam na nuvem. Isto significa que eles podem facilmente lidar com bancos de dados extremamente grandes e podem automaticamente obter mais recursos computacionais quando necessário.
-   SGBDs **No-processo** (*In-process*), como o *SQLite* ou *duckdb*, rodam inteiramente em seu computador. Eles são ótimos para trabalhar com conjunto de dados grandes nos quais você é o principal usuário.

## Conectando a um banco de dados

Para se conectar a um banco de dados pelo R, você usa um par de pacotes:

-   Você sempre usuará o DBI (**d***ata***b***ase* **i***nterface*), pois este fornece um conjunto de funções genéricas que se conectam com os bancos de dados, enviam dados, executam consultas SQL, etc.

-   Você também usará um pacote feito especificamente para o SGBD ao qual você está se conectando.

>     Este pacote traduz os comandos genéricos do DBI para as necessidades específicas do um dado SGBD.

Reviewer: ```suggestion
    Este pacote traduz os comandos genéricos do DBI para as necessidades específicas de um determinado SGBD.
```

Graças à magia do DBI, a única diferença entre usar o duckdb e outro SGBD é como você se conectará ao banco de dados.
Isto se torna muito bom para ensinar, pois você pode facilmente executar este código bem como pegar o que aprendeu e aplicar em outro lugar.

Connecting to duckdb is particularly simple because the defaults create a temporary database that is deleted when you quit R.
That's great for learning because it guarantees that you'll start from a clean slate every time you restart R:
Conectar ao duckdb é particularmente simples, pois por padrão, um banco de dados temporário é criado e depois removido ao sair do R.
Isto é muito bom para o aprendizado, pois garante que você iniciará de um estado limpo toda vez que reiniciar o R:

```{r}
con <- DBI::dbConnect(duckdb::duckdb())
```

duckdb is a high-performance database that's designed very much for the needs of a data scientist.
We use it here because it's very easy to get started with, but it's also capable of handling gigabytes of data with great speed.

> If you want to use duckdb for a real data analysis project, you'll also need to supply the `dbdir` argument to make a persistent database and tell duckdb where to save it.

Reviewer: Acho que "permanente" também fica estranho.

> If you want to use duckdb for a real data analysis project, you'll also need to supply the `dbdir` argument to make a persistent database and tell duckdb where to save it.

Author: My two cents: Persistente é um termo bastante comum no mundo dos bancos de dados, ele reforça a situação do dados persistir em dados e não em memória

> If you want to use duckdb for a real data analysis project, you'll also need to supply the `dbdir` argument to make a persistent database and tell duckdb where to save it.

Reviewer: Então vamos manter como persistente!

> Se você quiser usar o duckdb em um projeto de análise de dados real, será necessário incluir o argumento `dbdir` para um banco de dados permanente e dizer ao duckdb onde salvar.

Reviewer: ```suggestion
Se você quiser usar o duckdb em um projeto de análise de dados real, será necessário incluir o argumento `dbdir` para um banco de dados persistente e dizer ao duckdb onde salvar.
```

> Se você quiser usar o duckdb em um projeto de análise de dados real, você também precisará fornecer o argumento `dbdir` para persistir o banco de dados e dizer ao duckdb onde salvá-lo.

Reviewer: ```suggestion
Se você quiser usar o duckdb em um projeto de análise de dados real, será necessário incluir o argumento `dbdir` para um banco de dados permanente e dizer ao duckdb onde salvar.
```
Sugestão da Luana @Luaaalex.
Coloquei assim para facilitar a revisão, mas eu acho que essa tradução ainda não está refletindo bem...

Tô pensando, e para traduzir bem, acho que primeiro preciso entender o contexto e o que os autores querem dizer. Me parece que o ponto é indicar um diretório onde o bd será salvo, né? Ou entendi errado?

Reviewer: Qual a opinião de vocês @scopinho e @magnotairone ?

Author: oi @beatrizmilz , é isso mesmo. Por padrao o duckdb grava o db na memoria e nao 'persiste' o db em.disco a naonser que passe este argumento

> O exemplo mais simples de `dbWriteTable()` precisa de três argumentos: uma conexão ao banco de dados, o nome da tabela a ser criada no banco de dados e um *data frame* com os dados.

Reviewer 2: ```suggestion
O exemplo de uso mais simples de `dbWriteTable()` precisa de três argumentos: uma conexão ao banco de dados, o nome da tabela a ser criada no banco de dados e um *data frame* com os dados.
```

```{r}
dbWriteTable(con, "milhas", dados::milhas)
dbWriteTable(con, "diamante", dados::diamante)
```

Se você está usando o duckdb em um projeto real, recomendamos fortemente ler sobre `duckdb_read_csv()` e `duckdb_register_arrow()`.
Estas funções te dão formas poderosas e de alto desempenho para carregar dados diretamente ao duckdb sem ter que os carregar primeiro no R.
Nós também demonstraremos uma técnica útil para carregar vários arquivos em um banco de dados na @sec-save-database.

### O básico do DBI

Você pode confirmar se os dados foram carregados corretamente usando um par de funções do DBI: `dbListTables()` lista todas as tabelas da banco de dados[^databases-3] e `dbReadTable()` retorna o conteúdo de uma tabela.

[^databases-3]: Ao menos, todas as tabelas que você tem permissão para ver.

```{r}
dbListTables(con)

con |> 
  dbReadTable("diamante") |> 
  as_tibble()
```

> `dbReadTable()` retorna um `data.frame` portanto usamos `as_tibble()` para converter em um *tibble* para que imprima de forma mais bonita na tela.

Reviewer 2: ```suggestion
`dbReadTable()` retorna um `data.frame`, portanto usamos `as_tibble()` para converter em um *tibble* para que imprima de forma mais bonita na tela.
```

Se você já sabe SQL, você pode usar `dbGetQuery()` pata obter os resultados de uma consulta executada no banco de dados:

```{r}
sql <- "
  SELECT quilate, corte, transparencia, cor, preco 
  FROM diamante 
  WHERE preco > 15000
"
as_tibble(dbGetQuery(con, sql))
```

Se você nunca viu SQL antes, não se preocupe!
Em breve você aprenderá mais sobre isto.
Mas se você ler atentamente, você pode adivinhar que estamos selecionando cinco colunas do conjunto de dados diamante e todas as linhas onde o `preco` é maior que 15.000.

## O básico do dbplyr

Agora que você se conectou ao banco de dados e carregou alguns dados, você pode começar a aprender sobre dbplyr.

> dbplyr roda por tráz do dplyr (*dplyr* **backend**), isso significa que você continua a escrever códigos dplyr, mas o *backend* o executa de maneira diferente.

Reviewer 2: ```suggestion
dbplyr roda por tráz do dplyr, isso significa que você continua a escrever códigos dplyr, mas o *backend* o executa de maneira diferente.
```

Neste caso, o dbplyr traduz para SQL; existem outros *backends* incluindo [dtplyr](https://dtplyr.tidyverse.org) que traduz o código para [data.table](https://r-datatable.com) e [multidplyr](https://multidplyr.tidyverse.org) que executa o código em vários núcleos (*cores*).

Para usar dbplyr, primeiro você usa `tbl()` para criar um objeto que representa a tabela do banco de dados:

```{r}
diamantes_bd <- tbl(con, "diamante")
diamantes_bd
```

::: callout-note
Existem duas outras formas comuns de interagir com banco de dados.

> Primeira, muitas organizações possuem banco de dados muito grandes, então você precisa de alguma hierarquia para manter todas as tabelas organizadas.

Reviewer 2: ```suggestion
Primeiro, muitas organizações possuem banco de dados muito grandes, então você precisa de alguma hierarquia para manter todas as tabelas organizadas.
```

. Isto é o que dbplyr gera de uma tablea sem mudanças.
:

[^databases-4]: É estranho, mas dependendo do contexto, `SELECT` pode ser tanto uma declaração quanto uma cláusula.
    Para evitar esta confusão, iremos usar uma **consulta** `SELECT` ao invés de um declaração `SELECT`.

[^databases-5]: Ok, tecnicamente, somente o `SELECT` é necessário, já que você pode escrever consultas como `SELECT 1+1` para executar cálculos básicos.
    Mas se você quer trabalhar com dados (como você faz sempre!) você precisará da cláusula `FROM`.

```{r}
voos |> show_query()
avioes |> show_query()
```

`WHERE` e `ORDER BY` controlam quais linhas serão incluídas e como estarão ordenadas:

```{r}
voos |> 
  filter(destino == "IAH") |> 
  arrange(atraso_saida) |>
  show_query()
```

`GROUP BY` converte a consulta em uma sumarização, fazendo com que aconteça uma agregação:

```{r}
voos |> 
  group_by(destino) |> 
  summarize(atraso_saida = mean(atraso_saida, na.rm = TRUE)) |> 
  show_query()
```

Existem duas principais diferenças entre verbos dplyr e cláusulas SELECT:

> -   No SQL, variações de letras (maiúscula/minúscula) não faz diferença: você pode escrever `select`, `SELECT` ou até `SeLeCt`. Neste livro ficaremos com a convenção comum de escrever as palavras-chaves SQL usando letras maiúsculas para distinguir de nomes de tabelas ou variáveis.

Reviewer 2: ```suggestion
-   No SQL,  a maiúscula e minúscula não fazem diferença: você pode escrever `select`, `SELECT` ou até `SeLeCt`. Neste livro ficaremos com a convenção comum de escrever as palavras-chaves SQL usando letras maiúsculas para distinguir de nomes de tabelas ou variáveis.
```

> -   NO SQL, a ordem importa: você sempre deve escrever as declarações em ordem `SELECT`, `FROM`, `WHERE`, `GROUP BY`, `ORDER BY`. É um pouco confuso, pois esta ordem não é a mesma de como as declarações são realmente avaliadas com `FROM` em primeiro, depois `WHERE`, `GROUP BY`, `SELECT` e `ORDER BY`.

Reviewer 2: ```suggestion
-   No SQL, a ordem importa: você sempre deve escrever as declarações em ordem `SELECT`, `FROM`, `WHERE`, `GROUP BY`, `ORDER BY`. É um pouco confuso, pois esta ordem não é a mesma de como as declarações são realmente avaliadas com `FROM` em primeiro, depois `WHERE`, `GROUP BY`, `SELECT` e `ORDER BY`.
```

Não é perfeito, mas está melhorando continuamente e se você encontrar um problema, pode abrir um caso (*issue*) [no GitHub](https://github.com/tidyverse/dbplyr/issues/) para nos ajudar a melhorar.
:::

### SELECT

A cláusula `SELECT` é o motor das consultas e faz o mesmo trabalho que `select()`, `mutate()`, `rename()`, `relocate()` e, como você aprenderá na próxima seção, `summarize()`.

`select()`, `rename()` e `relocate()` tem uma tradução muito direta para o `SELECT` já que apenas afetam onde uma coluna aparece (e se aparece) assim como seu nome:

```{r}
avioes |> 
  select(codigo_cauda, tipo, fabricante, modelo, ano) |> 
  show_query()

avioes |> 
  select(codigo_cauda, tipo, fabricante, modelo, ano) |> 
  rename(ano_construcao = ano) |> 
  show_query()

avioes |> 
  select(codigo_cauda, tipo, fabricante, modelo, ano) |> 
  relocate(fabricante, modelo, .before = tipo) |> 
  show_query()
```

Este exemplo também mostra como SQL renomeia.
Na terminologia SQL, renomear é chamado de **aliasing** e é feito com `AS`.
Note que diferente de `mutate()`, o nome antigo vai ao lado esquerdo e o novo nome ao lado direito.

::: callout-note

> Nos exemplos acima, se tivessemos os nomes de `"ano"` e `"tipo"` em inglês (*year* e *type*), elas apareceriam entre aspas duplas.

Reviewer: ```suggestion
Nos exemplos acima, se tivéssemos os nomes de `"ano"` e `"tipo"`, elas apareceriam entre aspas duplas.
```

Note que nas funções de janela, a ordenação da informação se repete: a cláusula `ORDER BY` da consulta principal não se aplica automaticamente às funções de janela.

Outra função SQL importante é a `CASE WHEN`. É usada para traduzir as funções `if_else()` e `case_when()` que dplyr se inspirou diretamente.
Aqui estão alguns exemplos simples:

```{r}
voos |> 
  consulta_mutate(
    descricao = if_else(atraso_chegada > 0, "atrasado", "no horario")
  )
voos |> 
  consulta_mutate(
    descricao = 
      case_when(
        arr_delay < -5 ~ "adiantado", 
        arr_delay < 5 ~ "no horario",
        arr_delay >= 5 ~ "atrasado"
      )
  )
```

`CASE WHEN` também é usado para algumas funções que não tem tradução direta do R para o SQL.
Um bom exemplo disso é `cut()`:

```{r}
voos |> 
  consulta_mutate(
    descricao =  cut(
      atraso_chegada, 
      breaks = c(-Inf, -5, 5, Inf), 
      labels = c("adiantado", "no horario", "atrasado")
    )
  )
```

dbplyr também traduz funções comuns de manipulação de *string* e data/hora (*datetime*), as quais você pode aprender sobre na `vignette("translation-function", package = "dbplyr")`.

> As traduções do dbplyr certamente não são perfeitas, e ainda existem muitas funções do R que ainda não são traduzidas, mas faz um surpreendente bom trabalho em traduzir as funções que você usará na maior parte do tempo.

Reviewer 2: ```suggestion
As traduções do dbplyr certamente não são perfeitas, e ainda existem muitas funções do R que ainda não são traduzidas, mas o dbplyr faz um surpreendente bom trabalho em traduzir as funções que você usará na maior parte do tempo.
```

## Resumo

Neste capítulo você aprendeu como acessar dados de um banco de dados.

> Nós focamos no dbplyr, um "*backend*" do dplyr que permite que você escreva seu código dplyr que está familiarizado e o tenha automaticamente traduzido em SQL.

Reviewer 2: ```suggestion
Nós focamos no dbplyr, um "*backend*" do dplyr que permite que você escreva o código dplyr que está familiarizado e o tenha automaticamente traduzido em SQL.
```

Nós utilizamos esta tradução para te ensinar um pouco de SQL; é importante aprender SQL pois é *a* linguagem mais comumente utilizada para se trabalhar com dados e conhecê-la um pouco facilitrá sua comunicação com outras pessoas de dados que não usam o R.
Se você terminou este capítulo e gostaria de aprender mais sobre SQL.
Nós temos duas recomendações:

-   [*SQL for Data Scientists*](https://sqlfordatascientists.com) de Renée M. P. Teate é uma introdução ao SQL desenhada especificamente para as necessidades de cientistas de dados e inclui exemplos com dados altamente interconectados que você geralmente encontra em organizações reais.

> -   [*Practical SQL*](https://www.practicalsql.com) de Anthony DeBarros é escrito sob a perspectiva de uma pessoa jornalista de dados (cientista de dados especialista em contar atraentes histórias) e entra em mais detalhes sobre ter seus dados em um banco de dados e rodar seu próprio SGBD.

Reviewer 2: ```suggestion
-   [*Practical SQL*](https://www.practicalsql.com) de Anthony DeBarros é escrito sob a perspectiva de uma pessoa jornalista de dados (cientista de dados especialista em contar histórias atraentes) e entra em mais detalhes sobre ter seus dados em um banco de dados e rodar seu próprio SGBD.
```
